examples/tic-tac-toe/move/sources/owned.move

```diff
@@ -0,0 +1,326 @@
+/// running a centralized service that monitors marker placement events and
+/// responds to them, or it could be a 1-of-2 multisig address shared between
+/// the two players, as demonstrated in the demo app.
+///
+/// The `shared` module shows a variant of this game implemented using shared
+/// objects, which provides different trade-offs: Using shared objects is more
+/// expensive, however the implementation is more straightforward and each move
+/// only requires one transaction.
+module tic_tac_toe::owned {
+    use sui::event;
+    use sui::transfer::Receiving;
+
+    // === Object Types ===
+
+    /// The state of an active game of tic-tac-toe.
+    public struct Game has key, store {
+        id: UID,
+
+        /// Marks on the board.
+        board: vector<u8>,
+
+        /// The next turn to be played.
+        turn: u8,
+
+        /// The address expected to send moves on behalf of X.
+        x: address,
+
+        /// The address expected to send moves on behalf of O.
+        o: address,
+
+        /// Public key of the admin address.
+        admin: vector<u8>,
+    }
+
+    /// The player that the next turn is expected from is given a `TurnCap`.
+    public struct TurnCap has key {
+        id: UID,
+        game: ID,
+    }
+
+    /// A request to make a play -- only the player with the `TurnCap` can
+    /// create and send `Mark`s.
+    public struct Mark has key, store {
+        id: UID,
+        player: address,
+        row: u8,
+        col: u8,
+    }
+
+    /// An NFT representing a finished game. Sent to the winning player if there
+    /// is one, or to both players in the case of a draw.
+    public struct Trophy has key {
+        id: UID,
+
+        /// Whether the game was won or drawn.
+        status: u8,
+
+        /// The state of the board at the end of the game.
+        board: vector<u8>,
+
+        /// The number of turns played
+        turn: u8,
+
+        /// The other player (relative to the player who owns this Trophy).
+        other: address,
+    }
+
+    // === Event Types ===
+
+    public struct MarkSent has copy, drop {
+        game: ID,
+        mark: ID,
+    }
+
+    public struct GameEnd has copy, drop {
+        game: ID,
+    }
+
+    // === Constants ===
+
+    // Marks
+    const MARK__: u8 = 0;
+    const MARK_X: u8 = 1;
+    const MARK_O: u8 = 2;
+
+    // Trophy status
+    const TROPHY_NONE: u8 = 0;
+    const TROPHY_DRAW: u8 = 1;
+    const TROPHY_WIN: u8 = 2;
+
+    // === Errors ===
+
+    /// Move was for a position that doesn't exist on the board.
+    const EInvalidLocation: u64 = 0;
+
+    /// Game expected a move from another player.
+    const EWrongPlayer: u64 = 1;
+
+    /// Game has not reached an end condition.
+    const ENotFinished: u64 = 2;
+
+    /// Can't place a mark on a finished game.
+    const EAlreadyFinished: u64 = 3;
+
+    /// Game reached an end state that wasn't expected.
+    const EInvalidEndState: u64 = 4;
```

Comment:
I vote we refactor these to using clever errors

Reply:
Sounds good to me!

Reply:
Before:

> Error: MoveAbort(MoveLocation { module: ModuleId { address: d35ada3fa1a85b111c72865e5f60edc692aa98401e2f95a7449365bbb6d3f1a0, name: Identifier("shared") }, function: 1, instruction: 66, function_name: Some("place_mark") }, 9223372560841244680) in command 0

After:


> Error executing transaction 'D3PKdLRFhLP6MLjPAGcganYgacULoLxMAc63AZ9KU6PK': 1st command aborted within function '0xd35ada3fa1a85b111c72865e5f60edc692aa98401e2f95a7449365bbb6d3f1a0::shared::place_mark' at line 122. Aborted with 'EAlreadyFilled' -- 'Attempted to place a mark on a filled slot.'

Glorious!